File: src/routes/player.ts
import { Router, Request, Response } from "express"
import { groupBy, Months } from "../lib/misc.js"
import Player from "../models/player.js"
import Map from "../models/map.js"

const routes = Router()

async function players_overview_route(req: Request, res: Response) {
    const player = req.params.player
    const rankings = groupBy(await Player.rankings(player), "Category")
    if(!Object.keys(rankings).length)
        return res.render("pages/player/base.njk", { "search": true })

    const lastFinishes = await Player.lastFinishes(player, 10)
    const teamPartners = await Player.teamPartners(player, 10)
    const rankedPointsGraph = await Player.rankedpointsGraph(player)
    const pointsGraph = await Player.pointsGraph(player)
    const points = await Player.points(player)
    const page = req.params.type ?? "overview"
    
    const isMapper = (await Map.search({ mapper: player })).length
    const playerinfo = await Player.playerInfo(player)
    res.render("pages/player/overview/base.njk", { page, player, playerinfo, lastFinishes, teamPartners, points, rankedPointsGraph, pointsGraph, isMapper, "search": true })
}

routes.get("/player/json", players_json)
routes.get("/player/:player", players_overview_route)
routes.get("/player/:player/:type(overview)", players_overview_route)
routes.get("/player/:player/:type(overview)/finishes", async (req: Request, res: Response) => {
    const player = req.params.player
    const lastFinishes = await Player.lastFinishes(player, 100)
    const page = req.params.type

    const isMapper = (await Map.search({ mapper: player })).length
    const playerinfo = await Player.playerInfo(player)
    res.render("pages/player/overview/finishes.njk", { player, playerinfo, page, lastFinishes, isMapper, "search": true })
})
routes.get("/player/:player/:type(overview)/partners", async (req: Request, res: Response) => {
    const player = req.params.player
    const teamPartners = await Player.teamPartners(player, 10000)
    const page = req.params.type

    const isMapper = (await Map.search({ mapper: player })).length
    const playerinfo = await Player.playerInfo(player)
    res.render("pages/player/overview/partners.njk", { player, playerinfo, page, teamPartners, isMapper, "search": true })
})

async function players_finishes_route(req: Request, res: Response) {
    const player = req.params.player
    const rankings = groupBy(await Player.rankings(player), "Category")
    if(!Object.keys(rankings).length)
        return res.render("pages/player/base.njk", { "search": true })

    const playtime = await Player.playtime(player)
    const page = req.params.type ?? "overview"

    const isMapper = (await Map.search({ mapper: player })).length
    const playerinfo = await Player.playerInfo(player)
    res.render("pages/player/finishes.njk", { page, player, playerinfo, rankings, playtime, isMapper, "search": true })
}

routes.get("/player/:player/:type(finishes)", players_finishes_route)

async function players_activity_route(req: Request, res: Response) {
    const player = req.params.player
    const rankings = groupBy(await Player.rankings(player), "Category")
    if(!Object.keys(rankings).length)
        return res.render("pages/player/base.njk", { "search": true })

    const recentPlaytime = await Player.recentPlaytime(player, 11)
    const playtimeCategories = await Player.playtimeCategories(player)
    const playtimeGametypes = await Player.playtimeGametypes(player)
    const playtimeLocation = await Player.playtimeLocation(player)
    const playtimePerMonth = await Player.playtimePerMonth(player)
    const mostPlayedMaps = await Player.mostPlayedMaps(player, 11)
    const page = req.params.type

    const isMapper = (await Map.search({ mapper: player })).length
    const recentPlayerinfo = await Player.recentPlayerinfo(player, 5)
    const playerinfo = await Player.playerInfo(player)

    res.render("pages/player/activity/base.njk", { player, playerinfo, page, recentPlayerinfo, recentPlaytime, playtimeLocation, playtimeGametypes, 
        playtimePerMonth, playtimeCategories, Months, mostPlayedMaps, isMapper, "search": true })
}

routes.get("/player/:player/:type(activity)", players_activity_route)
routes.get("/player/:player/:type(activity)/mostplayed", async (req: Request, res: Response) => {
    const player = req.params.player
    const mostPlayedMaps = await Player.mostPlayedMaps(player, 10000)
    const page = req.params.type

    const isMapper = (await Map.search({ mapper: player })).length
    const playerinfo = await Player.playerInfo(player)
    res.render("pages/player/activity/mostplayed.njk", { player, playerinfo, page, mostPlayedMaps, isMapper, "search": true })
})

routes.get("/player/:player/:type(activity)/playerinfo", async (req: Request, res: Response) => {
    const player = req.params.player
    const recentPlayerinfo = await Player.recentPlayerinfo(player, 100)
    const page = req.params.type

    const isMapper = (await Map.search({ mapper: player })).length
    const playerinfo = await Player.playerInfo(player)

    res.render("pages/player/activity/playerinfo.njk", { player, playerinfo, page, recentPlayerinfo, isMapper, "search": true })
})

routes.get("/player/:player/:type(activity)/playtime", async (req: Request, res: Response) => {
    const player = req.params.player
    const recentPlaytime = await Player.recentPlaytime(player, 1000)
    const page = req.params.type

    const isMapper = (await Map.search({ mapper: player })).length
    const playerinfo = await Player.playerInfo(player)
    res.render("pages/player/activity/playtime.njk", { player, playerinfo, page, recentPlaytime, isMapper, "search": true })
})

routes.get("/player/:player/:type(playtime)/*", async (req: Request, res: Response) => {
    res.redirect(`/player/${req.params.player}/activity`)
})

async function players_rank1s_route(req: Request, res: Response) {
    const player = req.params.player
    const rankings = groupBy(await Player.rankings(player), "Category")
    if(!Object.keys(rankings).length)
        return res.render("pages/player/base.njk", { "search": true })

    const rank1sPartners = await Player.rank1sPartners(player, 10)
    const recentTop10s = await Player.recentTop10s(player, 10)
    const AmountOfTop10Placements = await Player.AmountOfTop10Placements(player)
    const allTop10s = await Player.allTop10s(player)
    const page = req.params.type
    
    const isMapper = (await Map.search({ mapper: player })).length
    const playerinfo = await Player.playerInfo(player)
    res.render("pages/player/rank1s/base.njk", { player, playerinfo, page, allTop10s, AmountOfTop10Placements, rank1sPartners, recentTop10s, isMapper, "search": true })
}

routes.get("/player/:player/:type(rank1s)/partners", async (req: Request, res: Response) => {
    const player = req.params.player
    const page = req.params.type

    const isMapper = (await Map.search({ mapper: player })).length
    const playerinfo = await Player.playerInfo(player)
    const rank1sPartners = await Player.rank1sPartners(player, 999)
    res.render("pages/player/rank1s/partners.njk", { rank1sPartners, page, player, playerinfo, isMapper, "search": true })
})

routes.get("/player/:player/:type(rank1s)", players_rank1s_route)

async function players_json(req: Request, res: Response) {
    const player = req.query.player
    const rankings = groupBy(await Player.rankings(player), "Category")
    if(!Object.keys(rankings).length)
        return res.status(404).json({
            "error": "player not found"
        })

    const rankedPointsGraph = await Player.rankedpointsGraph(player)
    const pointsGraph = await Player.pointsGraph(player)
    const playtime = await Player.playtime(player)
    const points = await Player.points(player)

    const recentPlaytime = await Player.recentPlaytime(player, 11)
    const playtimeCategories = await Player.playtimeCategories(player)
    const playtimeGametypes = await Player.playtimeGametypes(player)
    const playtimeLocation = await Player.playtimeLocation(player)
    const playtimePerMonth = await Player.playtimePerMonth(player)
    const mostPlayedMaps = await Player.mostPlayedMaps(player, 11)

    const rank1sPartners = await Player.rank1sPartners(player, 10)
    const recentTop10s = await Player.recentTop10s(player, 10)
    const AmountOfTop10Placements = await Player.AmountOfTop10Placements(player)
    const allTop10s = await Player.allTop10s(player)

    const isMapper = (await Map.search({ mapper: player })).length
    const recentPlayerinfo = await Player.recentPlayerinfo(player, 5)
    const playerinfo = await Player.playerInfo(player)
    return res.json({ player, playerinfo, points, rankings, isMapper, rankedPointsGraph, pointsGraph, playtime, recentPlayerinfo, recentPlaytime, playtimeCategories, playtimeGametypes, playtimeLocation, playtimePerMonth, mostPlayedMaps, allTop10s, AmountOfTop10Placements, rank1sPartners, recentTop10s })
}

export default routes

File: src/http/routes/player.rs
use crate::http::error::Error;
use crate::http::render;
use crate::http::AppState;
use crate::models::mapper::Mapper;
use crate::models::player::Finish;
use crate::models::player::Player;
use crate::utils::create_index_by_field;
use axum::extract::Path;
use axum::extract::Request;
use axum::extract::State;
use axum::http::StatusCode;
use axum::middleware;
use axum::middleware::Next;
use axum::response::Html;
use axum::response::IntoResponse;
use axum::response::Response;
use axum::{routing::get, Router};
use sqlx::Pool;
use sqlx::Postgres;
use tera::Context;

pub async fn player_context(db: &Pool<Postgres>, name: &str, page: &str) -> Result<Context, Error> {
    let profile = Player::get_profile(db, name).await?;
    let is_mapper = !Mapper::maps(db, name, Some(1)).await.unwrap().is_empty();
    let mut context = Context::new();
    context.insert("name", &name);
    context.insert("profile", &profile);
    context.insert("is_mapper", &is_mapper);
    context.insert("page", &page);

    Ok(context)
}

pub async fn player_overview(
    Path(name): Path<String>,
    State(state): State<AppState>,
) -> Result<Html<String>, Error> {
    let recent_finishes = Player::recent_finishes(&state.db, &name, Some(10)).await?;
    let favourite_teammates = Player::favourite_teammates(&state.db, &name, Some(10)).await?;

    let mut context = player_context(&state.db, &name, "overview").await?;
    context.insert("recent_finishes", &recent_finishes);
    context.insert("favourite_teammates", &favourite_teammates);

    render(state.template, "player/overview/overview.html", &context)
}

pub async fn player_overview_partners(
    Path(name): Path<String>,
    State(state): State<AppState>,
) -> Result<Html<String>, Error> {
    let favourite_teammates = Player::favourite_teammates(&state.db, &name, None).await?;

    let mut context = player_context(&state.db, &name, "overview").await?;
    context.insert("favourite_teammates", &favourite_teammates);

    render(
        state.template,
        "player/overview/favourite_partners.html",
        &context,
    )
}

pub async fn player_overview_finishes(
    Path(name): Path<String>,
    State(state): State<AppState>,
) -> Result<Html<String>, Error> {
    let recent_finishes = Player::recent_finishes(&state.db, &name, Some(100)).await?;

    let mut context = player_context(&state.db, &name, "overview").await?;
    context.insert("recent_finishes", &recent_finishes);

    render(
        state.template,
        "player/overview/recent_finishes.html",
        &context,
    )
}

pub async fn player_finishes(
    Path(name): Path<String>,
    State(state): State<AppState>,
) -> Result<Html<String>, Error> {
    let finishes = create_index_by_field::<Finish, String>(
        Player::finishes(&state.db, &name).await?,
        |finish| finish.map.server.clone(),
    );
    let points = Player::points(&state.points, &name);

    let mut context = player_context(&state.db, &name, "finishes").await?;
    context.insert("points", &points);
    context.insert("finishes", &finishes);

    render(state.template, "player/finishes/finishes.html", &context)
}

pub async fn player_activity(
    Path(name): Path<String>,
    State(state): State<AppState>,
) -> Result<Html<String>, Error> {
    let recent_activity = Player::recent_activity(&state.db, &name, Some(11)).await?;
    let recent_player_info = Player::recent_player_info(&state.db, &name, Some(5)).await?;
    let most_played_maps = Player::most_played_maps(&state.db, &name, Some(11)).await?;
    let most_played_gametypes = Player::most_played_gametypes(&state.db, &name, Some(15)).await?;
    let most_played_categories = Player::most_played_categories(&state.db, &name, Some(15)).await?;
    let most_played_locations = Player::most_played_locations(&state.db, &name, Some(15)).await?;
    let playtime_per_month = Player::playtime_per_month(&state.db, &name, Some(12)).await?;

    let mut context = player_context(&state.db, &name, "activity").await?;
    context.insert("recent_activity", &recent_activity);
    context.insert("recent_player_info", &recent_player_info);
    context.insert("most_played_maps", &most_played_maps);
    context.insert("most_played_gametypes", &most_played_gametypes);
    context.insert("most_played_categories", &most_played_categories);
    context.insert("most_played_locations", &most_played_locations);
    context.insert("playtime_per_month", &playtime_per_month);

    render(state.template, "player/activity/activity.html", &context)
}

pub async fn player_activity_playtime(
    Path(name): Path<String>,
    State(state): State<AppState>,
) -> Result<Html<String>, Error> {
    let recent_activity = Player::recent_activity(&state.db, &name, Some(1000)).await?;

    let mut context = player_context(&state.db, &name, "activity").await?;
    context.insert("recent_activity", &recent_activity);

    render(state.template, "player/activity/playtime.html", &context)
}

pub async fn player_activity_player_info(
    Path(name): Path<String>,
    State(state): State<AppState>,
) -> Result<Html<String>, Error> {
    let recent_player_info = Player::recent_player_info(&state.db, &name, Some(100)).await?;

    let mut context = player_context(&state.db, &name, "activity").await?;
    context.insert("recent_player_info", &recent_player_info);

    render(state.template, "player/activity/player_info.html", &context)
}

pub async fn player_activity_most_played_maps(
    Path(name): Path<String>,
    State(state): State<AppState>,
) -> Result<Html<String>, Error> {
    let most_played_maps = Player::most_played_maps(&state.db, &name, None).await?;

    let mut context = player_context(&state.db, &name, "activity").await?;
    context.insert("most_played_maps", &most_played_maps);

    render(
        state.template,
        "player/activity/most_played_maps.html",
        &context,
    )
}

pub async fn player_rank1s(
    Path(name): Path<String>,
    State(state): State<AppState>,
) -> Result<Html<String>, Error> {
    let favourite_rank1s_teammates =
        Player::favourite_rank1s_teammates(&state.db, &name, Some(10)).await?;
    let all_top_10s = Player::all_top_10s(&state.db, &name).await?;
    let recent_top_10s = Player::recent_top_10s(&state.db, &name, Some(9)).await?;

    let mut context = player_context(&state.db, &name, "rank1s").await?;
    context.insert("favourite_rank1s_teammates", &favourite_rank1s_teammates);
    context.insert("all_top_10s", &all_top_10s);
    context.insert("recent_top_10s", &recent_top_10s);

    render(state.template, "player/rank1s/rank1s.html", &context)
}

pub async fn player_rank1s_partners(
    Path(name): Path<String>,
    State(state): State<AppState>,
) -> Result<Html<String>, Error> {
    let favourite_rank1s_teammates =
        Player::favourite_rank1s_teammates(&state.db, &name, None).await?;

    let mut context = player_context(&state.db, &name, "rank1s").await?;
    context.insert("favourite_rank1s_teammates", &favourite_rank1s_teammates);

    render(state.template, "player/rank1s/partners.html", &context)
}

pub async fn player_middleware(
    Path(name): Path<String>,
    State(state): State<AppState>,
    // you can add more extractors here but the last
    // extractor must implement `FromRequest` which
    // `Request` does
    request: Request,
    next: Next,
) -> Response {
    match Player::get_profile(&state.db, &name).await.is_ok() {
        true => next.run(request).await,
        false => (
            StatusCode::NOT_FOUND,
            render(state.template, "player/404.html", &Context::new()).unwrap(),
        )
            .into_response(),
    }
}

pub fn router(state: AppState) -> Router<AppState> {
    Router::new()
        // Overview
        .route("/", get(player_overview))
        .route("/overview", get(player_overview))
        .route("/overview/finishes", get(player_overview_finishes))
        .route("/overview/partners", get(player_overview_partners))
        // Finishes
        .route("/finishes", get(player_finishes))
        // Activity
        .route("/activity", get(player_activity))
        .route("/activity/playtime", get(player_activity_playtime))
        .route("/activity/playerinfo", get(player_activity_player_info))
        .route(
            "/activity/mostplayed",
            get(player_activity_most_played_maps),
        )
        // Rank 1s
        .route("/rank1s", get(player_rank1s))
        .route("/rank1s/partners", get(player_rank1s_partners))
        .layer(middleware::from_fn_with_state(
            state.clone(),
            player_middleware,
        ))
}
